src/annotator/anchoring/xpath-util.js

```diff
@@ -0,0 +1,209 @@
+      path = '/' + elem.tagName.toLowerCase() + `[${idx}]` + path;
+      elem = elem.parentNode;
+    }
+    return path;
+  });
+  return paths.get();
+}
+
+/**
+ * A simple XPath evaluator using only standard DOM methods which can
+ * evaluate queries of the form /tag[index]/tag[index].
+ */
+export function simpleXPathPure(nodes, relativeRoot) {
+  function getPathSegment(node) {
+    const name = getNodeName(node);
+    const pos = getNodePosition(node);
+    return `${name}[${pos}]`;
+  }
+
+  let rootNode = relativeRoot;
+
+  function getPathTo(node) {
+    let xpath = '';
+    let elem = node;
+    while (elem !== rootNode) {
+      if (!elem) {
+        throw new Error(
+          'Called getPathTo on a node which was not a descendant of @rootNode. ' +
+            rootNode
+        );
+      }
+      xpath = getPathSegment(elem) + '/' + xpath;
+      elem = elem.parentNode;
+    }
+    xpath = '/' + xpath;
+    xpath = xpath.replace(/\/$/, '');
+    return xpath;
+  }
+  const paths = nodes.map((index, node) => {
+    return getPathTo(node);
+  });
+  return paths.get();
+}
+
+/**
+ * Flatten a nested array structure.
+ * TODO: use Array.prototype.flat and polyfill
+ */
+function flatten(array) {
+  const flatten = ary => {
+    let flat = [];
+    ary.forEach(el => {
+      if (el && Array.isArray(el)) {
+        flat = flat.concat(flatten(el));
+      } else {
+        flat = flat.concat(el);
+      }
+    });
+    return flat;
+  };
+  return flatten(array);
+}
+
+/**
+ * Finds all text nodes within the elements in the current collection.
+ * Returns a new jQuery collection of text nodes.
+ */
+export function getTextNodes(jq) {
+  const getTextNodes = node => {
+    if (node && node.nodeType !== Node.TEXT_NODE) {
+      let nodes = [];
+      if (node.nodeType !== Node.COMMENT_NODE) {
+        [...node.childNodes].forEach(child => {
+          nodes.push(getTextNodes(child));
+        });
+      }
+      return nodes;
+    } else {
+      return node;
+    }
+  };
+  return jq.map((index, node) => {
+    return flatten(getTextNodes(node));
+  });
+}
+
+/**
+ * Determine the last text node inside or before the given node.
+ */
+export function getLastTextNodeUpTo(n) {
+  switch (n.nodeType) {
+    case Node.TEXT_NODE:
+      return n; // We have found our text node.
+    case Node.ELEMENT_NODE:
+      // This is an element, we need to dig in
+      if (n.lastChild) {
+        // Does it have children at all?
+        const result = getLastTextNodeUpTo(n.lastChild);
+        if (result) {
+          return result;
+        }
+      }
+  }
+  // Could not find a text node in current node, go backwards
+  const prev = n.previousSibling;
+  if (prev) {
+    // eslint-disable-next-line no-unused-vars
+    return getLastTextNodeUpTo(prev);
+  } else {
+    return null;
+  }
+}
+
+/**
+ * Determine the first text node in or after the given jQuery node.
+ */
+export function getFirstTextNodeNotBefore(n) {
```

Comment:
```suggestion
export function getFirstTextNodeNotBefore(node) {
```

As above, avoid very short names except for local uses within a function where the meaning is obvious.